### PR TITLE
Winston port

### PR DESCRIPTION
**Pull request by Sezzary**
To test:
- Sounds.
- Object behaviour.

### CODE REVIEW

**Comment by Stranger1992** [comment]
https://github.com/TombEngine/Resources/raw/809ced3f88903d2fc3290f00080e80906f291305/Wad2%20Objects/Misc/TR3_Winston.wad2

You can find Winston (and his army gear) here ^^

**Comment by Stranger1992** [comment]
My main points are:

- The handling of the animations is odd and going to cause problems. All animations are taken from the `ID_ARMY_WINSTON` but what if you do not have him in the WAD?  Well the result is as shown below. Shooting `ID_WINSTON` will reduce his health to 0 and he will go haywire. I propose that the animations from `ID_ARMY_WINSTON` are exported and applied onto `ID_WINSTON`  and to set his mode (placid or army version) via an OCB.

https://github.com/MontyTRC89/TombEngine/assets/84292688/f2446ebf-e5cb-418c-a5f6-95fa8286949b

- The `ID_ARMY_WINSTON` should be a different index. There is a section for meshswaps.

**Review by Stranger1992 — CHANGES_REQUESTED** [review]
My main points are:

- The handling of the animations is odd and going to cause problems. All animations are taken from the `ID_ARMY_WINSTON` but what if you do not have him in the WAD?  Well the result is as shown below. Shooting `ID_WINSTON` will reduce his health to 0 and he will go haywire. I propose that the animations from `ID_ARMY_WINSTON` are exported and applied onto `ID_WINSTON`  and to set his mode (placid or army version) via an OCB.

https://github.com/MontyTRC89/TombEngine/assets/84292688/f2446ebf-e5cb-418c-a5f6-95fa8286949b

- The `ID_ARMY_WINSTON` should be a different index. There is a section for meshswaps.

**Review by Adngel — COMMENTED** [review]
Hi there. 

I took a view to the code and a test in game, there are two issues that called my attention.

- When I tested Winston, I took the butler object from TR3 house level (slot 361 because the 360 hasn't got all animations) and put it in the slot 298 for the TEN level. During the game, when I shoot at him, he gets frozen and  don't react anymore. (I think it may be when I "kill" him. At first sight I didn't saw the issue in the code. Maybe there a missing state transition in the item animations?

- I've seen in the code that there are some system sounds, could they be adapted to the animations? (For example, when Winton is using the army costume, his tray is empty, but it still play cups sounds because is called from code).

Apart of that,  the rest looks to be good, I hadn't got any other issue with Winston when I played with him in the test level.

File: TombEngine/Objects/game_object_ids.h [hunk 1]
```diff
@@ -229,11 +229,13 @@ enum GAME_OBJECT_ID : short
...
 	ID_BOSS_EXPLOSION_SHOCKWAVE,
 	ID_BOSS_EXPLOSION_RING,
-	ID_CLAW_MUTANT = 293,
+	ID_CLAW_MUTANT,
 	ID_WASP_MUTANT,
 	//295
 	ID_SKATEBOARD = 296,
 	ID_SKATEBOARD_KID,
+	ID_WINSTON,
+	ID_ARMY_WINSTON,
```
[the remark this answers is not in the document]
Comment (Nemoel-Tomo) [reply TombEngine/Objects/game_object_ids.h:238]: seems like this slot is only used to borrow animations from. 
line 177 in winston.cpp: SetAnimation(item, ID_ARMY_WINSTON, 16);
Normal Winston does not have all animations. 
Wouldn't it be better to make a new asset with all animations included so we only need one Winston slot?

**Comment by Nemoel-Tomo** [comment]
I have added a hitroutine to Winston. Now he can not explode.

**Comment by Stranger1992** [comment]
[TR3_Winston2.zip](https://github.com/MontyTRC89/TombEngine/files/13518661/TR3_Winston2.zip)
Here is a combined version of Winston. All ARMY_WINSTON anims have been removed and moved to the main WINSTON model.

File: TombEngine/Objects/game_object_ids.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (Stranger1992) [reply TombEngine/Objects/game_object_ids.h:238]: See below.

[the remark this answers is not in the document]
Comment (Adngel) [reply TombEngine/Objects/game_object_ids.h:238]: Yes, I think that's the best, provide 1 Winston with full animations, and sort the meshes changes with the same workflow of the other entities.

[the remark this answers is not in the document]
Comment (Nemoel-Tomo) [reply TombEngine/Objects/game_object_ids.h]: I have changed the behavior of Winston and I have deleted ID_ARMY_Winston.
Winston with OCB has army Winston behaviour. He can guard and has no cup sounds.
Winston without OCB  has normal Winston behaviour. He can not guard and has the cup sounds when walking.
Added ID_MESHSWAP_WINSTON_ARMY_OUTFIT    ID-Nr #1120

**Comment by Stranger1992** [comment]
Amended TRCatalog with ID_MESHSWAP_WINSTON_ARMY_OUTFIT and removed the moveable 299 [HERE](https://github.com/MontyTRC89/Tomb-Editor/commit/7e2766a12be42b4eaff24a0faee6b193d60b15dd)

**Comment by Stranger1992** [comment]
Winston's state and animation names have been added to TRCatalog. You can find the updated version that reflect these changes [here](https://github.com/TombEngine/Resources/raw/4c608407e6e66eeb1902216136a61ffe767adfac/Wad2%20Objects/Misc/TR3_Winston.wad2).

**Review by Stranger1992 — CHANGES_REQUESTED** [review]
Couple of issues here:

**OCB Change in Runtime**

If user is to change OCB of Winston during runtime: he will always fall backwards. This should be changed to his standing animation. Reason being is: as shown below I have a volume setup so that `OnEnter` his OCB changes to 1 and `OnExit` changes back to 0. However as he falls backwards, his OCB goes back to 0 and then he can't get back up.

PS: I also have a global event that says that if Winston's OCB is 1: his outfit will change:

``` lua

LevelFuncs.WinstonMeshSwap = function(activator)
        local numMeshes = 26 -- Total number of meshes to swap

        for i = 0, numMeshes - 1 do
            LevelFuncs.Engine.Node.SwapMoveableMesh("winston_278", i, TEN.Objects.ObjID.MESHSWAP_WINSTON_ARMY_OUTFIT, i)
        end
end

```


https://github.com/MontyTRC89/TombEngine/assets/84292688/a2d9afa5-0353-4040-b4bb-456513678ca7

**WInston does not face Lara when shooting.**

There are a couple of instances where, after being knocked down: Winston will not correctly rotate towards Lara with his tray up. 

https://github.com/MontyTRC89/TombEngine/assets/84292688/a418ee27-88c0-4851-baed-40df0c40c8df

**Hardcoded sound effects**

These should be removed and applied to the asset itself. This I can take of if needed.

**Pull request merged**